server/service/mdm.go

```diff
@@ -738,3 +738,110 @@ func (svc *Service) GetMDMCommandResults(ctx context.Context, commandUUID string
 	}
 	return results, nil
 }
+
+////////////////////////////////////////////////////////////////////////////////
+// GET /mdm/commands
+////////////////////////////////////////////////////////////////////////////////
+
+type listMDMCommandsRequest struct {
+	ListOptions fleet.ListOptions `url:"list_options"`
+}
+
+type listMDMCommandsResponse struct {
+	Results []*fleet.MDMCommand `json:"results"`
+	Err     error               `json:"error,omitempty"`
+}
+
+func (r listMDMCommandsResponse) error() error { return r.Err }
+
+func listMDMCommandsEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (errorer, error) {
+	req := request.(*listMDMCommandsRequest)
+	results, err := svc.ListMDMCommands(ctx, &fleet.MDMCommandListOptions{
+		ListOptions: req.ListOptions,
+	})
+	if err != nil {
+		return listMDMCommandsResponse{
+			Err: err,
+		}, nil
+	}
+
+	return listMDMCommandsResponse{
+		Results: results,
+	}, nil
+}
+
+func (svc *Service) ListMDMCommands(ctx context.Context, opts *fleet.MDMCommandListOptions) ([]*fleet.MDMCommand, error) {
+	// first, authorize that the user has the right to list hosts
+	if err := svc.authz.Authorize(ctx, &fleet.Host{}, fleet.ActionList); err != nil {
+		return nil, ctxerr.Wrap(ctx, err)
+	}
+
+	vc, ok := viewer.FromContext(ctx)
+	if !ok {
+		return nil, fleet.ErrNoContext
+	}
+
+	// get the list of commands so we know what hosts (and therefore what teams)
+	// we're dealing with. Including the observers as they are allowed to view
+	// MDM Apple commands.
+	results, err := svc.ds.ListMDMCommands(ctx, fleet.TeamFilter{
+		User:            vc.User,
+		IncludeObserver: true,
+	}, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	// collect the different team IDs and verify that the user has access to view
+	// commands on all affected teams, do not assume that ListMDMCommands
+	// only returned hosts that the user is authorized to view the command
+	// results of (that is, always verify with our rego authz policy).
+	teamIDs := make(map[uint]bool)
+	for _, res := range results {
+		var id uint
+		if res.TeamID != nil {
+			id = *res.TeamID
+		}
+		teamIDs[id] = true
+	}
+
+	// instead of returning an authz error if the user is not authorized for a
+	// team, we remove those commands from the results (as we want to return
+	// whatever the user is allowed to see). Since this can only be done after
+	// retrieving the list of commands, this may result in returning less results
+	// than requested, but it's ok - it's expected that the results retrieved
+	// from the datastore will all be authorized for the user.
```

Comment:
Do we include tests somewhere for returning only a subset of the commands results based on team auth?